keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -638,24 +640,165 @@ always dispatched immediately.  Following is how the other requests
+  // `priority` MUST BE between 0 and 10000, inclusive, and
+  // SHOULD BE greater than zero.
+  // If it is zero then, for the sake of a smooth transition from the time
+  // before this field existed, this level will be treated as if its `priority`
+  // is the average of the `matchingPrecedence` of the FlowSchema objects
+  // that reference this level.
+  // +optional
+  Priority int32
+}
 ```
 
-where SCL is the apiserver's concurrency limit and ACS(l) is the
-AssuredConcurrencyShares for priority level l.
+Prior to the introduction of borrowing, the `assuredConcurrencyShares`
+field had two meanings that amounted to the same thing: the total
+shares of the level, and the non-borrowable shares of the level.
+While it is somewhat unnatural to keep the meaning of "total shares"
+for a field named "assured" shares, rolling out the new behavior into
+existing systems will be more continuous if we keep the meaning of
+"total shares" for the existing field.  In the next version we should
+rename the `AssuredConcurrencyShares` to `NominalConcurrencyShares`.
+
+Consider rolling the borrowing behavior into a pre-existing cluster
+that did not have borrowing.  In particular, suppose that the
+administrators of the cluster have scripts/clients that maintain some
+out-of-tree PriorityLevelConfiguration objects; these, naturally, do
+not specify a value for the `priority` field.  The default behavior
+for `priority` is designed to do something more natural and convenient
+than have them all collide at some fixed number.
+
+Borrowing is done on the basis of the current situation, with no
+consideration of opportunity cost, no further rationing according to
+shares (just obeying the concurrency limits as outlined above), and no
+pre-emption when the situation changes.
+
+Whenever a request is dispatched, it takes all its seats from one
+priority level --- either the one referenced by the request's
+FlowSchema or a lower priority level.
+
+Borrowing is further limited by a practical consideration: we do not
+want mutual exclusion among all priority levels in the dispatching
+code.  Aside from borrowing, dispatching from one priority level is
+done independently from dispatching at another --- using a distinct
+`sync.Mutex` for each priority level.  There _is_ a global
+`sync.RWMutex`, but queuing and dispatching only requires locking that
+for reading (which is not mutually exclusive); write locking is done
+only while reconfiguring.  Borrowing is allowed in just one direction:
+higher may borrow from lower.  A higher priority level may actively
+borrow from a lower one but a lower level does not actively lend to a
+higher one (see below).  In this design, working on dispatching one
+request requires holding at most two priority level mutexes at any
+given moment.  To avoid deadlock, there is a strict ordering on
+acquisition of those mutexes (namely, decreasing logical priority).
+
+We could take a complementary approach, in which lower actively lends
```

Comment:
This is a good observation of the current state.  I'm wondering how we would choose to configure priority levels after this.  Would we choose to reserve large portions for the cluster critical services and then have a low concurrency for workload low and simply let it borrow from the higher levels when there was space available.

cc @wojtek-t @tkashem for comment.

Reply:
That's actually a good point, I was thinking about that at some point but then forget.

This is actually a problem in the current design, where we're explicitly saying in lines 733-734 that we're not considering borrowing from higher levels to lower levels.
So David - what you're suggesting here will not happen in the current design.

The way I would like it to work is indeed how you're saying: "we reserve the large portion for cluster stuff that we may potentially need, and then let other to borrow from that if those are underutilized". But that implies that cluster stuff is lowest level, which is counter-intuitive...
OTOH, I really don't want to allocate a small fraction from cluster and let it borrow from other levels in case cluster is hugely utilized - that would be counter-productive.


Strawman (although I just came up with this idea and have no idea if it makes sense yet): what if the `priority` only defines the ordering (then maybe we don't need it even?) but in fact every level can borrow from any other?

Reply:
What we would need then though is more like rule:
"if there are requests at a given PL that could consume a capacity from it, then these have precendence over any other requests that would like to borrow the capacity"...

With the above, by giving good chunk of capacity to cluster stuff we ensure that it will have capacity when we need to and it's also borrowable.

And FWIW - I think we don't need the "priority" field at all (we can define ordering based on namespace/name or sth) and from that perspective it would be closer to what Daniel was proposing (i.e. we conceptually give capacity to others, it's just we still dispatch only from one level, not from the whole bag?)

Reply:
We have four proposals on the table: this PR, its complement, #3389, and #3391.